src/ashley/api/views.py

```diff
@@ -0,0 +1,132 @@
+PermissionRequiredMixin: BasePermissionRequiredMixin = get_class(
+    "forum_permission.viewmixins", "PermissionRequiredMixin"
+)
+logger = logging.getLogger(__name__)
+
+
+class UserApiView(
+    mixins.ListModelMixin, mixins.UpdateModelMixin, viewsets.GenericViewSet
+):  # pylint: disable=too-many-ancestors
+    """API requests to list moderators for current LTIContext"""
+
+    # Add permission to check if user has the right permission and can manage moderators
+    permission_classes = [ManageModeratorPermission]
+    serializer_class = UserSerializer
+    ordering = "public_username"
+
+    def get_queryset(self):
+        context = get_current_lti_session(self.request)
+        role = self.request.GET.get("role", False)
+
+        if role == _FORUM_ROLE_MODERATOR:
+            # Gets all the moderators for the current LTIContext, exclude instructors
+            # make sure user has base group from this context"""
+            return (
+                User.objects.filter(
+                    is_active=True,
+                    groups__name=context.get_group_role_name(_FORUM_ROLE_MODERATOR),
+                )
+                .filter(groups__name=context.base_group_name)
+                .exclude(
+                    groups__name=context.get_group_role_name(_FORUM_ROLE_INSTRUCTOR)
+                )
+                .order_by(self.ordering)
+            )
+        if role == f"!{_FORUM_ROLE_MODERATOR}":
+            # Gets all the users that aren't moderators for the current LTIContext and exclude
+            # instructors."""
+            return (
+                User.objects.filter(
+                    is_active=True,
+                    groups__name=context.base_group_name,
+                )
+                .exclude(
+                    groups__name=context.get_group_role_name(_FORUM_ROLE_INSTRUCTOR)
+                )
+                .exclude(
+                    groups__name=context.get_group_role_name(_FORUM_ROLE_MODERATOR)
+                )
+                .order_by(self.ordering)
+            )
+        if role == _FORUM_ROLE_INSTRUCTOR:
+            # Gets all the instructor for the current LTIContext.
+            return User.objects.filter(
+                is_active=True,
+                groups__name=context.get_group_role_name(_FORUM_ROLE_INSTRUCTOR),
+            ).order_by(self.ordering)
+
+        if role == _FORUM_ROLE_STUDENT:
+            # Gets all the students for the current LTIContext.
+            return User.objects.filter(
+                is_active=True,
+                groups__name=context.get_group_role_name(_FORUM_ROLE_STUDENT),
+            ).order_by(self.ordering)
+
+        # If no role is specified, filter all the users for the current LTIContext
+        return (
+            User.objects.filter(
+                is_active=True,
+                groups__name=context.base_group_name,
+            )
+            .order_by(self.ordering)
+            .distinct()
+        )
```

Comment:
I think this function can be factored to avoid repeating treatments on each case (e.g. filtering active_users and sorting).

Also, I think that you are trying to handle an exhaustive list of cases that cannot predicted due to the fact that LTI role names are not standardized and depends of the LTI Consumer implementation.

I suggest the following rewrite of the function, with these notable changes:
 - I removed the if `if role == _FORUM_ROLE_INSTRUCTOR` and `if role == _FORUM_ROLE_STUDENT`
 - I generalized the use of the `!` prefix to exclude a role from the results to every possible roles

```suggestion
    def get_queryset(self):
        context = get_current_lti_session(self.request)
        role = self.request.GET.get("role", None)

        # Select all active users who are part of the current LTIContext
        query = User.objects.filter(
            is_active=True,
            groups__name=context.base_group_name,
        )

        # If a role is specified and start with !, we exclude its members from the results
        if role is not None and role.startswith("!"):
            query = query.exclude(groups__name=context.get_group_role_name(role[1:]))
        # If a role is specified, we select only users having this role inside the LTIContext
        elif role:
            query = query.filter(groups__name=context.get_group_role_name(role))

        # When the role moderator (or !moderator) is specified, we want to exclude the
        # users having the instructor role from the results.
        if role in [_FORUM_ROLE_MODERATOR, f"!{_FORUM_ROLE_MODERATOR}"]:
            query = query.exclude(groups__name=context.get_group_role_name(_FORUM_ROLE_INSTRUCTOR))

        return query.order_by(self.ordering)
```

Reply:
much better !